### PR TITLE
DVC Blog

### PR DESCRIPTION
**Pull request by max-hoffman**
New branch to avoid confusion w/ metaflow.

Original comments in https://github.com/dolthub/dolt-integrations/pull/7

### CODE REVIEW

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+1. Ease of access
+2. Remotes and sharing
+3. Lineage
+4. Reproducibility
+
+# Background
+
+The machine learning tooling space is changing quickly, budding hundreds of
+competing and overlapping narratives of how the future will
+look in data-land. Some projects, like Blue River, expand
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: Simplify

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+the capabilities of specific verticals, like autonomous tractors. Others
+like Weights and Biases support specific stages in the ML
+lifecycle, like training and result tracking.
+
+"Data versioning and reproducibility" has a deceivingly simple scope
+that masks the variety and difficulty of data capture.
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: Simplify:
> ...sound simple enough, but are in fact complicated in practice.

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+
+Consider two versioning tools: Dolt and Data Version Control.
+
+Dolt is an SQL database with Git-versioning semantics.
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:25]: I prefer "Git-like version control features"

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+Tabular datasets in Dolt gain the powers of production databases
+(scalability, schema reliability, query language, logging) and Git
+(row-level commits, diffs, and merges, etc).
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: This is a good sentence. Maybe remove parens and make it two?

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+
+Data Version Control (DVC) uses workflow files to support team
+collaboration of Git source code and remote object stores.
+
+Because Dolt is an unopinionated data store, and DVC is an opinionated
+workflow manager, we believe the two support one-another.
+We built an integration to show how this works in practice.
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:35]: Excellent, addresses the core piece of feedback I had for the first version

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+We chose this to tutorial to leverage Dolt and DVC’s strengths:
+1. DVC documents workflows by recording commands and the metadata of files
+   synced to remote file-systems.
+2. Dolt is a versioned SQL database with novel features for managing
+   strictly-typed tabular data.
+
+We divide input and output dependencies between Dolt and DVC according
+to those strengths:
+
+TODO diagram
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:68]: I am on it

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+sample](https://www.tensorflow.org/tutorials/keras/classification_files/output_Ko-uzOufSCSe_0.png)
+
+## Setup
+
+Install dolt
+```bash
+sudo bash -c 'curl -L https://github.com/dolthub/dolt/releases/latest/download/install.sh | sudo bash'
+```
+
+Clone and install DVC's `dolt-integration` branch:
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:117]: It this branch on their repository, or a fork?

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+> dvc init
+```
+
+Finally, let's download training code adapted from [Tensorflow
+documentation](https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/classification.ipynb).
+```bash
+> dvc get git@github.com:dolthub/dolt-integrations.git dolt_integrations/dvc/examples/fashion_mnist/main.py
+```
+
+Please reference the Tensorflow docs for tensorflow, keras and other
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: capitalize tensorflow and keras

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+
+	git add data/.gitignore data/images.dvc
+```
+
+If we pulled files from an existing DVC repo the add step would be performed automatically.
+
+### Dolt
+
+Getting a [Dolt
+table ](https://www.dolthub.com/repositories/max-hoffman/mnist)
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: space after table

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+In [2]: with gzip.open("scratch/mnist/train-labels-idx1-ubyte.gz","r") as f:
+   ...:     f.read(8) # first two bytes are padded zeros
+   ...:     buf = f.read()
+   ...:     labels = np.frombuffer(buf, dtype=np.uint8).astype(np.int64)
+
+In [2]: os.makedirs("mnist", exist_ok=True)
+   ...: dolt.Dolt.init("mnist")
+   ...: db = dolt.Dolt("mnist")
+   ...: n = len(labels)
+   ...: dolt.write_columns(db, "labels", dict(row_id=range(n), train=[True]*n, label=labels), primary_key=["row_id"])
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:193]: You can commit from this with `commit=True`

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 11]
```diff
@@ -0,0 +1,610 @@
...
+> dvc push data/images -r dvc_remote
+```
+
+If a co-worker added new data, we would first use Git to reference
+new DVC metadata:
+```bash
+> git pull
+```
+
+then download remotes into our cache:
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: Capitlize

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 12]
```diff
@@ -0,0 +1,610 @@
...
+```
+
+Remove and restore the database with both branches:
+```bash
+> rm -rf data/labels
+> dvc pull data/labels -r dolt_remote
+1 file fetched
+```
+
+If you accidentally switch up the mnist images and labels, and fall back
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: Capitalize MNIST

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+
+## Lineage
+
+Before running our workflow, lets setup one more database for the
+prediction results:
+```bash
+> mkdir -p data/predictions
+> cd data/predictions
+> dolt init
+```
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:286]: Use `&&` to make this a one-liner?

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 14]
```diff
@@ -0,0 +1,610 @@
...
+Epoch 9/10
+1875/1875 [==============================] - 2s 1ms/step - loss: 1.2764 - accuracy: 0.6741
+Epoch 10/10
+1875/1875 [==============================] - 2s 998us/step - loss: 1.2752 - accuracy: 0.6725
+313/313 - 0s - loss: 1.7931 - accuracy: 0.5620
+Updating lock file 'dvc.lock'
+
+To track the changes with git, run:
+
+	git add dvc.lock
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:430]: code block

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 15]
```diff
@@ -0,0 +1,610 @@
...
+diff --dolt a/labels b/labels
+--- a/labels @ h6a5mj4rle852jj2133dtl0ofi6hdsqf
++++ b/labels @ bcvu38tkg94vmvua2qi1c9rpsg4tktpn
+7,038 Rows Unmodified (10.05%)
+0 Rows Added (0.00%)
+0 Rows Deleted (0.00%)
+62,962 Rows Modified (89.95%)
+62,962 Cells Modified (29.98%)
+(70,000 Entries vs 70,000 Entries)
+```
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:455]: This is cool, maybe lead with this?

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+    count(*) as row_number
+  from t
+  group by commit_hash"
++-----------------------+------------+----------------------------------+
+| sum(correct)/count(*) | row_number | commit_hash                      |
++-----------------------+------------+----------------------------------+
+| 0.6217                | 10000      | vq01vp8o1fsdq95eo4sutmppfhovejfv |
+| 0.5352                | 10000      | foo013e1hgej8oa0282bv1ecb9280mlh |
++-----------------------+------------+----------------------------------+
+```
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:486]: More amazing content, maybe lead with it?

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 17]
```diff
@@ -0,0 +1,610 @@
...
+One instance where we might want to access the commit history is
+`dolt blame`’ing a data changes. For example, identifying who added
+certain rows to the fashion branch:
+```bash
+dolt blame fashion labels | head -n 5
++--------+----------------------------+-----------------+------------------------------+----------------------------------+
+| ROW_ID | COMMIT MSG                 | AUTHOR          | TIME                         | COMMIT                           |
++--------+----------------------------+-----------------+------------------------------+----------------------------------+
+| 14648  | Add Fashion labels         | Bojack Horseman | Wed Mar 31 16:20:50 PDT 2021 | 5fblpjp5neurvsfp989s6ea9lt01vd2q |
+| 27051  | Add Fashion labels         | Bojack Horseman | Wed Mar 31 16:20:50 PDT 2021 | 5fblpjp5neurvsfp989s6ea9lt01vd2q |
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:518]: Maybe do this from SQL?

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 18]
```diff
@@ -0,0 +1,610 @@
...
+outs:
+- md5: tkcao72e0upe2umb2kokartgivd9keqc-pinisl8m4tfdecb8iikqhicta61nou9s.dolt
+  size: 1465033
+  path: labels
+  cache: false
+```
+
+Dolt's three  heads tell us 1) the last commit, 2) the hash if
+all files in our working directory were committed now
+, 3) the ongoing index/staging commit hash.
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: Dolt doesn't track files though? I assume you mean DVC?

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md [hunk 19]
```diff
@@ -0,0 +1,610 @@
...
+> dolt status
+On branch master
+Changes not staged for commit:
+  (use "dolt add <table>" to update what will be committed)
+  (use "dolt checkout <table>" to discard changes in working directory)
+	modified:       summary
+
+> dvc add data/predictions
+Adding...
+ERROR: unexpected error - Dolt status is not clean; commit a reproducible state before adding.
```
Comment (oscarbatori) [inline dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:570]: This is good

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md (hunk 7 above)
Comment (max-hoffman) [reply dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:117]: ah true

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md (hunk 13 above)
Comment (max-hoffman) [reply dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:286]: does it improve readability? i've been erring towards readable > functional, similar to the write + commit as two steps

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md (hunk 14 above)
Comment (max-hoffman) [reply dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:430]: this stdout is legit, from dvc

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md (hunk 18 above)
Comment (max-hoffman) [reply dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md]: right, rows not files

File: dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md (hunk 17 above)
Comment (max-hoffman) [reply dolt_integrations/dvc/examples/fashion-mnist/dolt-dvc-demo.md:518]: yeah that'd be better, i was a bit clumsy with that table when i first wrote this but i'll take another look
